### PR TITLE
mpOpenAPI-4.0: implement validation

### LINKED ISSUE #28860: Implement OpenAPI document validation on MP OpenAPI 4.0
Older `mpOpenAPI` features produce OpenAPI v3.0 documents and we validate the document after it's created and report warnings for any validation issues.

`mpOpenAPI-4.0` can produce OpenAPI v3.1 documents, which have different validation rules. We need to implement and test the new validation rules to be used when generating an OpenAPI v3.1 document.

`mpOpenAPI-4.0` can be configured to create OpenAPI v3.0 documents. In this case we'd want to use the old 3.0 validation rules, but the model API is incompatible in places so we may not be able to reuse the existing validation code directly.

### PR DESCRIPTION
**Pull request by Azquelt**
- [x] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [x] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [x] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

Fixes #28860

### CODE REVIEW

**Comment by Azquelt** [comment]
#libby

**Comment by Azquelt** [comment]
#build
#spawn.fullfat.buckets=com.ibm.ws.microprofile.openapi.1.1_fat_tck,com.ibm.ws.microprofile.openapi_fat,com.ibm.ws.microprofile.openapi_fat_tck,com.ibm.ws.openapi.ui_fat,io.openliberty.microprofile.openapi.2.0.internal_fat,io.openliberty.microprofile.openapi.2.0.internal_fat_tck,io.openliberty.microprofile.openapi.3.0.internal_fat_tck,io.openliberty.microprofile.openapi.3.1.internal_fat_tck,io.openliberty.microprofile.openapi.4.0.internal_fat,io.openliberty.microprofile.openapi.4.0.internal_fat_tck,io.openliberty.microprofile.openapi.ui.internal_fat

**Comment by Azquelt** [comment]
> **The following files contain a string indicating discriminating content:**

This should now be fixed.

File: dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/validation/ValidationMessages.nlsprops [hunk 1]
```diff
@@ -64,6 +64,8 @@ nullValueInMap=The map specifies an invalid value for the \"{0}\" key. A map sho
 
 openAPIVersionInvalid=The OpenAPI Object must contain a valid OpenAPI specification version. The \"{0}\" value specified for the OpenAPI specification version is not valid
 openAPITagIsNotUnique=The OpenAPI Object must contain unique tag names. The \"{0}\" tag name is not unique
+#3.1
+openAPIMissingRequiredFields=The OpenAPI Object must contain at least one of the \"paths\", \"components\" or \"webhooks\" properties
```
Comment (is273) [inline dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/validation/ValidationMessages.nlsprops]: The OpenAPI Object must contain at least one of the \"paths\", \"components\", or \"webhooks\" properties

File: dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/validation/ValidationMessages.nlsprops [hunk 2]
```diff
@@ -112,9 +115,21 @@ schemaMultipleOfNotGreaterThanZero=The Schema Object must have the \"multipleOf\
 schemaPropertyLessThanZero=The \"{0}\" property of the Schema Object must be greater than or equal to zero
 schemaTypeDoesNotMatchProperty=The \"{0}\" property is not appropriate for the Schema Object of \"{1}\" type
 
+#3.1
+schemaPropertyWrongType=The \"{0}\" property of the Schema Object must be of type \"{1}\"
```
Comment (is273) [inline dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/validation/ValidationMessages.nlsprops]: The \"{0}\" property of the Schema Object must be a \"{1}\" type

File: dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/validation/ValidationMessages.nlsprops [hunk 3]
```diff
@@ -112,9 +115,21 @@ schemaMultipleOfNotGreaterThanZero=The Schema Object must have the \"multipleOf\
...
+schemaPropertyEmptyArray=The \"{0}\" property of the Schema Object is an empty array
+schemaPropertyNotUnique=The elements of the \"{0}\" property of the Schema Object are not unique
+schemaInvalidType=The \"type\" property of the Schema Object must only contain values which are one of (\"null\", \"boolean\", \"object\", \"array\", \"number\", \"integer\", \"string\")
```
Comment (is273) [inline dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/validation/ValidationMessages.nlsprops]: The \"type\" property of the Schema Object must contain values that are one of (\"null\", \"boolean\", \"object\", \"array\", \"number\", \"integer\", or \"string\")

File: dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/validation/ValidationMessages.nlsprops [hunk 4]
```diff
@@ -112,9 +115,21 @@ schemaMultipleOfNotGreaterThanZero=The Schema Object must have the \"multipleOf\
...
+schemaCollectionElementWrongType=The \"{0}\" property of the Schema Object contains a value which is not of type \"{1}\"
```
Comment (is273) [inline dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/validation/ValidationMessages.nlsprops]: The \"{0}\" property of the Schema Object contains a value that is not a \"{1}\" type

**Comment by Azquelt** [comment]
Fixed OSGi services not present during unit testing.

#build
#spawn.fullfat.buckets=com.ibm.ws.microprofile.openapi.1.1_fat_tck,com.ibm.ws.microprofile.openapi_fat,com.ibm.ws.microprofile.openapi_fat_tck,com.ibm.ws.openapi.ui_fat,io.openliberty.microprofile.openapi.2.0.internal_fat,io.openliberty.microprofile.openapi.2.0.internal_fat_tck,io.openliberty.microprofile.openapi.3.0.internal_fat_tck,io.openliberty.microprofile.openapi.3.1.internal_fat_tck,io.openliberty.microprofile.openapi.4.0.internal_fat,io.openliberty.microprofile.openapi.4.0.internal_fat_tck,io.openliberty.microprofile.openapi.ui.internal_fat

**Comment by Azquelt** [comment]
Personal build looks good
#libby

File: dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/validation/ValidationMessages.nlsprops (hunk 4 above)
Comment (helyarp) [inline dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/validation/ValidationMessages.nlsprops]: Replace "which" with "that" or add a comma before "which"

**Review by helyarp — COMMENTED** [review]
Suggested a minor change

File: dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/validation/ValidationMessages.nlsprops (hunk 2 above)
Comment (Azquelt) [reply dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/validation/ValidationMessages.nlsprops]: Is this still ok if the `{1}` variable starts with a vowel?

The values used for `{1}` in the code at the moment are "string", "number", "integer", "array", "object", "boolean", and "schema"

File: dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/validation/ValidationMessages.nlsprops (hunk 4 above)
Comment (Azquelt) [reply dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/validation/ValidationMessages.nlsprops]: `{1}` may start with a vowel here too (same question as the above comment)

File: dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/validation/ValidationMessages.nlsprops (hunk 2 above)
Comment (Azquelt) [reply dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/validation/ValidationMessages.nlsprops]: We decided on `The type of the "{0}" property of the Schema Object must be "{1}"`

File: dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/validation/ValidationMessages.nlsprops (hunk 4 above)
Comment (Azquelt) [reply dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/validation/ValidationMessages.nlsprops]: We decided on `The "{0}" property of the Schema Object contains a value whose type is not "{1}"`

**Comment by Azquelt** [comment]
Message review changes made.

#libby
#build
#spawn.fullfat.buckets=com.ibm.ws.microprofile.openapi.1.1_fat_tck,com.ibm.ws.microprofile.openapi_fat,com.ibm.ws.microprofile.openapi_fat_tck,com.ibm.ws.openapi.ui_fat,io.openliberty.microprofile.openapi.2.0.internal_fat,io.openliberty.microprofile.openapi.2.0.internal_fat_tck,io.openliberty.microprofile.openapi.3.0.internal_fat_tck,io.openliberty.microprofile.openapi.3.1.internal_fat_tck,io.openliberty.microprofile.openapi.4.0.internal_fat,io.openliberty.microprofile.openapi.4.0.internal_fat_tck,io.openliberty.microprofile.openapi.ui.internal_fat

**Review by is273 — COMMENTED** [review]
LGTM

**Comment by Azquelt** [comment]
Missed a message change so need to re-request the build

#build
#spawn.fullfat.buckets=com.ibm.ws.microprofile.openapi.1.1_fat_tck,com.ibm.ws.microprofile.openapi_fat,com.ibm.ws.microprofile.openapi_fat_tck,com.ibm.ws.openapi.ui_fat,io.openliberty.microprofile.openapi.2.0.internal_fat,io.openliberty.microprofile.openapi.2.0.internal_fat_tck,io.openliberty.microprofile.openapi.3.0.internal_fat_tck,io.openliberty.microprofile.openapi.3.1.internal_fat_tck,io.openliberty.microprofile.openapi.4.0.internal_fat,io.openliberty.microprofile.openapi.4.0.internal_fat_tck,io.openliberty.microprofile.openapi.ui.internal_fat

**Comment by Azquelt** [comment]
Personal build looks good

File: dev/com.ibm.ws.microprofile.openapi_fat/fat/src/com/ibm/ws/microprofile/openapi/validation/fat/OpenAPIValidationTestFive.java
Comment (Yas-Franco) [inline dev/com.ibm.ws.microprofile.openapi_fat/fat/src/com/ibm/ws/microprofile/openapi/validation/fat/OpenAPIValidationTestFive.java:1]: The copyright headers are missing from all files in this folder

File: dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/validation/ExampleValidator.java
Comment (Yas-Franco) [inline dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/validation/ExampleValidator.java:1]: Copyright header date needs updating

File: dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/validation/HeaderValidator.java
Comment (Yas-Franco) [inline dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/validation/HeaderValidator.java:1]: Copyright header date needs updating

File: dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/validation/LinkValidator.java
Comment (Yas-Franco) [inline dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/validation/LinkValidator.java:1]: Copyright header date needs updating

File: dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/validation/ParameterValidator.java
Comment (Yas-Franco) [inline dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/validation/ParameterValidator.java:1]: Copyright header date needs updating

File: dev/io.openliberty.microprofile.openapi.2.0.internal_test/src/io/openliberty/microprofile/openapi20/test/utils/TestValidationContextHelper.java
Comment (Yas-Franco) [inline dev/io.openliberty.microprofile.openapi.2.0.internal_test/src/io/openliberty/microprofile/openapi20/test/utils/TestValidationContextHelper.java:1]: Copyright header date needs updating

File: dev/io.openliberty.microprofile.openapi.2.0.internal_test/src/io/openliberty/microprofile/openapi20/validation/test/OpenAPIValidatorTest.java
Comment (Yas-Franco) [inline dev/io.openliberty.microprofile.openapi.2.0.internal_test/src/io/openliberty/microprofile/openapi20/validation/test/OpenAPIValidatorTest.java:1]: Copyright header date needs updating

File: dev/io.openliberty.microprofile.openapi.2.0.internal_test/src/io/openliberty/microprofile/openapi20/validation/test/OperationValidatorTest.java
Comment (Yas-Franco) [inline dev/io.openliberty.microprofile.openapi.2.0.internal_test/src/io/openliberty/microprofile/openapi20/validation/test/OperationValidatorTest.java:1]: Copyright header date needs updating

File: dev/io.openliberty.microprofile.openapi.4.0.internal_fat/bnd.bnd
Comment (Yas-Franco) [inline dev/io.openliberty.microprofile.openapi.4.0.internal_fat/bnd.bnd:1]: Copyright header date needs updating

**Review by Yas-Franco — DISMISSED** [review]

File: dev/com.ibm.ws.microprofile.openapi_fat/fat/src/com/ibm/ws/microprofile/openapi/validation/fat/OpenAPIValidationTestFive.java
Comment (Azquelt) [reply dev/com.ibm.ws.microprofile.openapi_fat/fat/src/com/ibm/ws/microprofile/openapi/validation/fat/OpenAPIValidationTestFive.java:1]: These should be fixed in a follow up, we'd need to go back and look through the history to find out when each file was created.

Opened #29733 for this.

File: dev/io.openliberty.microprofile.openapi.2.0.internal_test/src/io/openliberty/microprofile/openapi20/test/utils/TestValidationContextHelper.java
Comment (Azquelt) [reply dev/io.openliberty.microprofile.openapi.2.0.internal_test/src/io/openliberty/microprofile/openapi20/test/utils/TestValidationContextHelper.java:1]: I don't think these changes are large enough to warrant updating the copyright date. This is just the knock-on effects of doing a refactor in another file.

**Comment by Azquelt** [comment]
Squashed the message reviews and updated the copyright headers.

#libby
#build

**Review by Yas-Franco — APPROVED** [review]

**Comment by Azquelt** [comment]
Errors in the build do not appear to related to this change.

**Pull request merged**